File: rust/src/stream/proto.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum MarketTypeMsg {
    PumpFun,
    PumpSwap,
    MeteoraDbc,
    MeteoraDammV2,
    RaydiumLaunchpad,
    RaydiumCpmm,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PumpFunContextMsg {}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PumpSwapContextMsg {
    pub pool: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub global_config: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MeteoraDbcContextMsg {
    pub pool: String,
    pub config: String,
    pub quote_mint: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MeteoraDammV2ContextMsg {
    pub pool: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RaydiumLaunchpadContextMsg {
    pub pool: String,
    pub config: String,
    pub platform: String,
    pub quote_mint: String,
    pub user_quote_account: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RaydiumCpmmContextMsg {
    pub pool: String,
    pub config: String,
    pub quote_mint: String,
    pub user_quote_account: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MarketContextMsg {
    pub market_type: MarketTypeMsg,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pumpfun: Option<PumpFunContextMsg>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pumpswap: Option<PumpSwapContextMsg>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meteora_dbc: Option<MeteoraDbcContextMsg>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meteora_damm_v2: Option<MeteoraDammV2ContextMsg>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raydium_launchpad: Option<RaydiumLaunchpadContextMsg>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raydium_cpmm: Option<RaydiumCpmmContextMsg>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StrategyConfigMsg {
    pub target_profit_pct: f64,
    pub stop_loss_pct: f64,
    pub deadline_timeout_sec: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum SessionModeMsg {
    FullWalletWatch,
    Integration,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct IntegrationContextMsg {
    #[serde(flatten)]
    pub fields: BTreeMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct LimitsMsg {
    pub hi_capacity: u32,
    pub pnl_flush_ms: u64,
    pub max_positions_per_session: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    Ping {
        client_time_ms: u64,
    },
    Configure {
        wallet_pubkey: String,
        mode: SessionModeMsg,
        strategy: StrategyConfigMsg,
        #[serde(skip_serializing_if = "Option::is_none")]
        integration_context: Option<IntegrationContextMsg>,
    },
    UpdateStrategy {
        strategy: StrategyConfigMsg,
    },
    ClosePosition {
        position_id: u64,
    },
    RequestExitSignal {
        position_id: u64,
        #[serde(skip_serializing_if = "Option::is_none")]
        slippage_bps: Option<u16>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    HelloOk {
        session_id: u64,
        server_time_ms: u64,
        limits: LimitsMsg,
        #[serde(skip_serializing_if = "Option::is_none")]
        mode: Option<SessionModeMsg>,
    },
    Pong {
        server_time_ms: u64,
    },
    Error {
        code: String,
        message: String,
    },
    PnlUpdate {
        position_id: u64,
        profit_units: i64,
        proceeds_units: u64,
        server_time_ms: u64,
    },
    BalanceUpdate {
        wallet_pubkey: String,
        mint: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        token_account: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        token_program: Option<String>,
        tokens: u64,
        slot: u64,
    },
    PositionOpened {
        position_id: u64,
        wallet_pubkey: String,
        mint: String,
        token_account: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        token_program: Option<String>,
        tokens: u64,
        entry_quote_units: u64,
        #[serde(skip_serializing_if = "Option::is_none")]
        market_context: Option<MarketContextMsg>,
        slot: u64,
    },
    PositionClosed {
        position_id: u64,
        wallet_pubkey: String,
        mint: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        token_account: Option<String>,
        reason: String,
        slot: u64,
    },
    ExitSignalWithTx {
        session_id: u64,
        position_id: u64,
        wallet_pubkey: String,
        mint: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        token_account: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        token_program: Option<String>,
        position_tokens: u64,
        profit_units: i64,
        reason: String,
        triggered_at_ms: u64,
        #[serde(skip_serializing_if = "Option::is_none")]
        market_context: Option<MarketContextMsg>,
        unsigned_tx_b64: String,
    },
}

impl ClientMessage {
    pub fn from_text(text: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(text)
    }

    pub fn to_text(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
}

impl ServerMessage {
    pub fn from_text(text: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(text)
    }

    pub fn to_text(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn round_trip<T>(value: T)
    where
        T: Serialize + for<'de> Deserialize<'de> + PartialEq + std::fmt::Debug,
    {
        let json = serde_json::to_string(&value).expect("serialize");
        let decoded: T = serde_json::from_str(&json).expect("deserialize");
        assert_eq!(value, decoded);
    }

    fn base_context() -> MarketContextMsg {
        MarketContextMsg {
            market_type: MarketTypeMsg::PumpFun,
            pumpfun: None,
            pumpswap: None,
            meteora_dbc: None,
            meteora_damm_v2: None,
            raydium_launchpad: None,
            raydium_cpmm: None,
        }
    }

    #[test]
    fn market_context_round_trip_pumpfun() {
        let mut ctx = base_context();
        ctx.market_type = MarketTypeMsg::PumpFun;
        ctx.pumpfun = Some(PumpFunContextMsg {});
        round_trip(ctx);
    }

    #[test]
    fn market_context_round_trip_pumpswap() {
        let mut ctx = base_context();
        ctx.market_type = MarketTypeMsg::PumpSwap;
        ctx.pumpswap = Some(PumpSwapContextMsg {
            pool: "11111111111111111111111111111111".to_string(),
            global_config: Some("22222222222222222222222222222222".to_string()),
        });
        round_trip(ctx);
    }

    #[test]
    fn market_context_round_trip_meteora_dbc() {
        let mut ctx = base_context();
        ctx.market_type = MarketTypeMsg::MeteoraDbc;
        ctx.meteora_dbc = Some(MeteoraDbcContextMsg {
            pool: "11111111111111111111111111111111".to_string(),
            config: "22222222222222222222222222222222".to_string(),
            quote_mint: "33333333333333333333333333333333".to_string(),
        });
        round_trip(ctx);
    }

    #[test]
    fn market_context_round_trip_meteora_damm_v2() {
        let mut ctx = base_context();
        ctx.market_type = MarketTypeMsg::MeteoraDammV2;
        ctx.meteora_damm_v2 = Some(MeteoraDammV2ContextMsg {
            pool: "11111111111111111111111111111111".to_string(),
        });
        round_trip(ctx);
    }

    #[test]
    fn market_context_round_trip_raydium_launchpad() {
        let mut ctx = base_context();
        ctx.market_type = MarketTypeMsg::RaydiumLaunchpad;
        ctx.raydium_launchpad = Some(RaydiumLaunchpadContextMsg {
            pool: "11111111111111111111111111111111".to_string(),
            config: "22222222222222222222222222222222".to_string(),
            platform: "33333333333333333333333333333333".to_string(),
            quote_mint: "44444444444444444444444444444444".to_string(),
            user_quote_account: "55555555555555555555555555555555".to_string(),
        });
        round_trip(ctx);
    }

    #[test]
    fn market_context_round_trip_raydium_cpmm() {
        let mut ctx = base_context();
        ctx.market_type = MarketTypeMsg::RaydiumCpmm;
        ctx.raydium_cpmm = Some(RaydiumCpmmContextMsg {
            pool: "11111111111111111111111111111111".to_string(),
            config: "22222222222222222222222222222222".to_string(),
            quote_mint: "33333333333333333333333333333333".to_string(),
            user_quote_account: "44444444444444444444444444444444".to_string(),
        });
        round_trip(ctx);
    }

    #[test]
    fn client_configure_round_trip() {
        let mut integration_fields = BTreeMap::new();
        integration_fields.insert(
            "provider".to_string(),
            serde_json::Value::String("mobile".to_string()),
        );

        let msg = ClientMessage::Configure {
            wallet_pubkey: "11111111111111111111111111111111".to_string(),
            mode: SessionModeMsg::Integration,
            strategy: StrategyConfigMsg {
                target_profit_pct: 5.0,
                stop_loss_pct: 1.5,
                deadline_timeout_sec: 45,
            },
            integration_context: Some(IntegrationContextMsg {
                fields: integration_fields,
            }),
        };

        round_trip(msg.clone());
        let encoded = msg.to_text().expect("encode");
        let decoded = ClientMessage::from_text(&encoded).expect("decode");
        assert_eq!(decoded, msg);
    }

    #[test]
    fn server_hello_ok_round_trip() {
        let msg = ServerMessage::HelloOk {
            session_id: 42,
            server_time_ms: 1700000000000,
            limits: LimitsMsg {
                hi_capacity: 256,
                pnl_flush_ms: 100,
                max_positions_per_session: 256,
            },
            mode: Some(SessionModeMsg::FullWalletWatch),
        };

        round_trip(msg.clone());
        let encoded = msg.to_text().expect("encode");
        let decoded = ServerMessage::from_text(&encoded).expect("decode");
        assert_eq!(decoded, msg);
    }

    #[test]
    fn exit_signal_with_tx_round_trip() {
        let ctx = MarketContextMsg {
            market_type: MarketTypeMsg::RaydiumCpmm,
            pumpfun: None,
            pumpswap: None,
            meteora_dbc: None,
            meteora_damm_v2: None,
            raydium_launchpad: None,
            raydium_cpmm: Some(RaydiumCpmmContextMsg {
                pool: "11111111111111111111111111111111".to_string(),
                config: "22222222222222222222222222222222".to_string(),
                quote_mint: "33333333333333333333333333333333".to_string(),
                user_quote_account: "44444444444444444444444444444444".to_string(),
            }),
        };

        let msg = ServerMessage::ExitSignalWithTx {
            session_id: 7,
            position_id: 8,
            wallet_pubkey: "55555555555555555555555555555555".to_string(),
            mint: "11111111111111111111111111111111".to_string(),
            token_account: Some("22222222222222222222222222222222".to_string()),
            token_program: None,
            position_tokens: 10,
            profit_units: 5,
            reason: "tp".to_string(),
            triggered_at_ms: 123,
            market_context: Some(ctx),
            unsigned_tx_b64: "dGVzdA==".to_string(),
        };

        round_trip(msg);
    }

    #[test]
    fn pnl_update_round_trip() {
        let msg = ServerMessage::PnlUpdate {
            position_id: 5,
            profit_units: 12,
            proceeds_units: 34,
            server_time_ms: 999,
        };
        round_trip(msg);
    }
}

File: python/src/lasersell_sdk/stream/proto.py
"""Stream protocol messages ported from lasersell-stream-proto."""

from __future__ import annotations

import json
import math
from typing import Any, Literal, NotRequired, Required, TypedDict, cast

MarketTypeMsg = Literal[
    "pump_fun",
    "pump_swap",
    "meteora_dbc",
    "meteora_damm_v2",
    "raydium_launchpad",
    "raydium_cpmm",
]


class PumpFunContextMsg(TypedDict):
    pass


class PumpSwapContextMsg(TypedDict, total=False):
    pool: Required[str]
    global_config: NotRequired[str]


class MeteoraDbcContextMsg(TypedDict):
    pool: str
    config: str
    quote_mint: str


class MeteoraDammV2ContextMsg(TypedDict):
    pool: str


class RaydiumLaunchpadContextMsg(TypedDict):
    pool: str
    config: str
    platform: str
    quote_mint: str
    user_quote_account: str


class RaydiumCpmmContextMsg(TypedDict):
    pool: str
    config: str
    quote_mint: str
    user_quote_account: str


class MarketContextMsg(TypedDict, total=False):
    market_type: Required[MarketTypeMsg]
    pumpfun: NotRequired[PumpFunContextMsg]
    pumpswap: NotRequired[PumpSwapContextMsg]
    meteora_dbc: NotRequired[MeteoraDbcContextMsg]
    meteora_damm_v2: NotRequired[MeteoraDammV2ContextMsg]
    raydium_launchpad: NotRequired[RaydiumLaunchpadContextMsg]
    raydium_cpmm: NotRequired[RaydiumCpmmContextMsg]


class StrategyConfigMsg(TypedDict, total=False):
    target_profit_pct: Required[float]
    stop_loss_pct: Required[float]
    trailing_stop_pct: NotRequired[float]
    sell_on_graduation: NotRequired[bool]


class LimitsMsg(TypedDict):
    hi_capacity: int
    pnl_flush_ms: int
    max_positions_per_session: int
    max_wallets_per_session: int
    max_positions_per_wallet: int
    max_sessions_per_api_key: int


class PingClientMessage(TypedDict):
    type: Literal["ping"]
    client_time_ms: int


class ConfigureClientMessage(TypedDict):
    type: Literal["configure"]
    wallet_pubkeys: list[str]
    strategy: StrategyConfigMsg


class UpdateStrategyClientMessage(TypedDict):
    type: Literal["update_strategy"]
    strategy: StrategyConfigMsg


class ClosePositionClientMessage(TypedDict, total=False):
    type: Required[Literal["close_position"]]
    position_id: NotRequired[int]
    token_account: NotRequired[str]


class RequestExitSignalClientMessage(TypedDict, total=False):
    type: Required[Literal["request_exit_signal"]]
    position_id: NotRequired[int]
    token_account: NotRequired[str]
    slippage_bps: NotRequired[int]


class UpdateWalletsClientMessage(TypedDict):
    type: Literal["update_wallets"]
    wallet_pubkeys: list[str]


ClientMessage = (
    PingClientMessage
    | ConfigureClientMessage
    | UpdateStrategyClientMessage
    | ClosePositionClientMessage
    | RequestExitSignalClientMessage
    | UpdateWalletsClientMessage
)


class HelloOkServerMessage(TypedDict):
    type: Literal["hello_ok"]
    session_id: int
    server_time_ms: int
    limits: LimitsMsg


class PongServerMessage(TypedDict):
    type: Literal["pong"]
    server_time_ms: int


class ErrorServerMessage(TypedDict):
    type: Literal["error"]
    code: str
    message: str


class PnlUpdateServerMessage(TypedDict):
    type: Literal["pnl_update"]
    position_id: int
    profit_units: int
    proceeds_units: int
    server_time_ms: int


class BalanceUpdateServerMessage(TypedDict, total=False):
    type: Required[Literal["balance_update"]]
    wallet_pubkey: Required[str]
    mint: Required[str]
    token_account: NotRequired[str]
    token_program: NotRequired[str]
    tokens: Required[int]
    slot: Required[int]


class PositionOpenedServerMessage(TypedDict, total=False):
    type: Required[Literal["position_opened"]]
    position_id: Required[int]
    wallet_pubkey: Required[str]
    mint: Required[str]
    token_account: Required[str]
    token_program: NotRequired[str]
    tokens: Required[int]
    entry_quote_units: Required[int]
    market_context: NotRequired[MarketContextMsg]
    slot: Required[int]


class PositionClosedServerMessage(TypedDict, total=False):
    type: Required[Literal["position_closed"]]
    position_id: Required[int]
    wallet_pubkey: Required[str]
    mint: Required[str]
    token_account: NotRequired[str]
    reason: Required[str]
    slot: Required[int]


class ExitSignalWithTxServerMessage(TypedDict, total=False):
    type: Required[Literal["exit_signal_with_tx"]]
    session_id: Required[int]
    position_id: Required[int]
    wallet_pubkey: Required[str]
    mint: Required[str]
    token_account: NotRequired[str]
    token_program: NotRequired[str]
    position_tokens: Required[int]
    profit_units: Required[int]
    reason: Required[str]
    triggered_at_ms: Required[int]
    market_context: NotRequired[MarketContextMsg]
    unsigned_tx_b64: Required[str]


ServerMessage = (
    HelloOkServerMessage
    | PongServerMessage
    | ErrorServerMessage
    | PnlUpdateServerMessage
    | BalanceUpdateServerMessage
    | PositionOpenedServerMessage
    | PositionClosedServerMessage
    | ExitSignalWithTxServerMessage
)


def client_message_from_text(text: str) -> ClientMessage:
    """Parses JSON text into a validated client message."""

    try:
        parsed = json.loads(text)
    except json.JSONDecodeError as exc:
        raise ValueError(f"decode client message JSON: {exc}") from exc
    return client_message_from_obj(parsed)


def client_message_from_obj(value: object) -> ClientMessage:
    """Parses a Python object into a validated client message."""

    obj = _as_record(value, "client message")
    message_type = _as_string(obj.get("type"), "client message.type")

    if message_type == "ping":
        return {
            "type": "ping",
            "client_time_ms": _as_int(obj.get("client_time_ms"), "ping.client_time_ms"),
        }

    if message_type == "configure":
        return {
            "type": "configure",
            "wallet_pubkeys": _parse_wallet_pubkeys(obj),
            "strategy": _parse_strategy_config(obj.get("strategy")),
        }

    if message_type == "update_strategy":
        return {
            "type": "update_strategy",
            "strategy": _parse_strategy_config(obj.get("strategy")),
        }

    if message_type == "close_position":
        message: ClosePositionClientMessage = {"type": "close_position"}
        _set_optional_int(
            message,
            "position_id",
            obj.get("position_id"),
            "close_position.position_id",
        )
        _set_optional_str(
            message,
            "token_account",
            obj.get("token_account"),
            "close_position.token_account",
        )
        return message

    if message_type in ("sell_now", "request_exit_signal"):
        message = cast(RequestExitSignalClientMessage, {"type": "request_exit_signal"})
        _set_optional_int(
            message,
            "position_id",
            obj.get("position_id"),
            "request_exit_signal.position_id",
        )
        _set_optional_str(
            message,
            "token_account",
            obj.get("token_account"),
            "request_exit_signal.token_account",
        )
        _set_optional_int(
            message,
            "slippage_bps",
            obj.get("slippage_bps"),
            "request_exit_signal.slippage_bps",
        )
        return message

    if message_type == "update_wallets":
        return {
            "type": "update_wallets",
            "wallet_pubkeys": _parse_wallet_pubkeys(obj),
        }

    raise ValueError(f"unsupported client message type: {message_type}")


def client_message_to_text(message: ClientMessage) -> str:
    """Serializes a client message to JSON text."""

    return json.dumps(message, separators=(",", ":"))


def server_message_from_text(text: str) -> ServerMessage:
    """Parses JSON text into a validated server message."""

    try:
        parsed = json.loads(text)
    except json.JSONDecodeError as exc:
        raise ValueError(f"decode server message JSON: {exc}") from exc
    return server_message_from_obj(parsed)


def server_message_from_obj(value: object) -> ServerMessage:
    """Parses a Python object into a validated server message."""

    obj = _as_record(value, "server message")
    message_type = _as_string(obj.get("type"), "server message.type")

    if message_type == "hello_ok":
        return {
            "type": "hello_ok",
            "session_id": _as_int(obj.get("session_id"), "hello_ok.session_id"),
            "server_time_ms": _as_int(obj.get("server_time_ms"), "hello_ok.server_time_ms"),
            "limits": _parse_limits(obj.get("limits")),
        }

    if message_type == "pong":
        return {
            "type": "pong",
            "server_time_ms": _as_int(obj.get("server_time_ms"), "pong.server_time_ms"),
        }

    if message_type == "error":
        return {
            "type": "error",
            "code": _as_string(obj.get("code"), "error.code"),
            "message": _as_string(obj.get("message"), "error.message"),
        }

    if message_type == "pnl_update":
        return {
            "type": "pnl_update",
            "position_id": _as_int(obj.get("position_id"), "pnl_update.position_id"),
            "profit_units": _as_int(obj.get("profit_units"), "pnl_update.profit_units"),
            "proceeds_units": _as_int(obj.get("proceeds_units"), "pnl_update.proceeds_units"),
            "server_time_ms": _as_int(obj.get("server_time_ms"), "pnl_update.server_time_ms"),
        }

    if message_type == "balance_update":
        message = cast(
            BalanceUpdateServerMessage,
            {
                "type": "balance_update",
                "wallet_pubkey": _as_string(
                    obj.get("wallet_pubkey"),
                    "balance_update.wallet_pubkey",
                ),
                "mint": _as_string(obj.get("mint"), "balance_update.mint"),
                "tokens": _as_int(obj.get("tokens"), "balance_update.tokens"),
                "slot": _as_int(obj.get("slot"), "balance_update.slot"),
            },
        )
        _set_optional_str(
            message,
            "token_account",
            obj.get("token_account"),
            "balance_update.token_account",
        )
        _set_optional_str(
            message,
            "token_program",
            obj.get("token_program"),
            "balance_update.token_program",
        )
        return message

    if message_type == "position_opened":
        message = cast(
            PositionOpenedServerMessage,
            {
                "type": "position_opened",
                "position_id": _as_int(obj.get("position_id"), "position_opened.position_id"),
                "wallet_pubkey": _as_string(
                    obj.get("wallet_pubkey"),
                    "position_opened.wallet_pubkey",
                ),
                "mint": _as_string(obj.get("mint"), "position_opened.mint"),
                "token_account": _as_string(
                    obj.get("token_account"),
                    "position_opened.token_account",
                ),
                "tokens": _as_int(obj.get("tokens"), "position_opened.tokens"),
                "entry_quote_units": _as_int(
                    obj.get("entry_quote_units"),
                    "position_opened.entry_quote_units",
                ),
                "slot": _as_int(obj.get("slot"), "position_opened.slot"),
            },
        )
        _set_optional_str(
            message,
            "token_program",
            obj.get("token_program"),
            "position_opened.token_program",
        )
        _set_optional_market_context(
            message,
            "market_context",
            obj.get("market_context"),
            "position_opened.market_context",
        )
        return message

    if message_type == "position_closed":
        message = cast(
            PositionClosedServerMessage,
            {
                "type": "position_closed",
                "position_id": _as_int(obj.get("position_id"), "position_closed.position_id"),
                "wallet_pubkey": _as_string(
                    obj.get("wallet_pubkey"),
                    "position_closed.wallet_pubkey",
                ),
                "mint": _as_string(obj.get("mint"), "position_closed.mint"),
                "reason": _as_string(obj.get("reason"), "position_closed.reason"),
                "slot": _as_int(obj.get("slot"), "position_closed.slot"),
            },
        )
        _set_optional_str(
            message,
            "token_account",
            obj.get("token_account"),
            "position_closed.token_account",
        )
        return message

    if message_type == "exit_signal_with_tx":
        message = cast(
            ExitSignalWithTxServerMessage,
            {
                "type": "exit_signal_with_tx",
                "session_id": _as_int(obj.get("session_id"), "exit_signal_with_tx.session_id"),
                "position_id": _as_int(
                    obj.get("position_id"),
                    "exit_signal_with_tx.position_id",
                ),
                "wallet_pubkey": _as_string(
                    obj.get("wallet_pubkey"),
                    "exit_signal_with_tx.wallet_pubkey",
                ),
                "mint": _as_string(obj.get("mint"), "exit_signal_with_tx.mint"),
                "position_tokens": _as_int(
                    obj.get("position_tokens"),
                    "exit_signal_with_tx.position_tokens",
                ),
                "profit_units": _as_int(
                    obj.get("profit_units"),
                    "exit_signal_with_tx.profit_units",
                ),
                "reason": _as_string(obj.get("reason"), "exit_signal_with_tx.reason"),
                "triggered_at_ms": _as_int(
                    obj.get("triggered_at_ms"),
                    "exit_signal_with_tx.triggered_at_ms",
                ),
                "unsigned_tx_b64": _as_string(
                    obj.get("unsigned_tx_b64"),
                    "exit_signal_with_tx.unsigned_tx_b64",
                ),
            },
        )
        _set_optional_str(
            message,
            "token_account",
            obj.get("token_account"),
            "exit_signal_with_tx.token_account",
        )
        _set_optional_str(
            message,
            "token_program",
            obj.get("token_program"),
            "exit_signal_with_tx.token_program",
        )
        _set_optional_market_context(
            message,
            "market_context",
            obj.get("market_context"),
            "exit_signal_with_tx.market_context",
        )
        return message

    raise ValueError(f"unsupported server message type: {message_type}")


def server_message_to_text(message: ServerMessage) -> str:
    """Serializes a server message to JSON text."""

    return json.dumps(message, separators=(",", ":"))


def _parse_wallet_pubkeys(obj: dict[str, object]) -> list[str]:
    wallets = obj.get("wallet_pubkeys", obj.get("wallet_pubkey"))

    if isinstance(wallets, str):
        return [wallets]

    if isinstance(wallets, list):
        return [_as_string(item, f"configure.wallet_pubkeys[{idx}]") for idx, item in enumerate(wallets)]

    raise ValueError(
        "configure.wallet_pubkeys must be a string or string[] (legacy wallet_pubkey is supported)"
    )


def _parse_strategy_config(value: object) -> StrategyConfigMsg:
    obj = _as_record(value, "strategy")

    result: StrategyConfigMsg = {
        "target_profit_pct": _as_float(obj.get("target_profit_pct"), "strategy.target_profit_pct"),
        "stop_loss_pct": _as_float(obj.get("stop_loss_pct"), "strategy.stop_loss_pct"),
    }

    trailing_raw = obj.get("trailing_stop_pct")
    if trailing_raw is not None:
        result["trailing_stop_pct"] = _as_float(trailing_raw, "strategy.trailing_stop_pct")

    sell_on_graduation_raw = obj.get("sell_on_graduation")
    if sell_on_graduation_raw is not None:
        if not isinstance(sell_on_graduation_raw, bool):
            raise ValueError("strategy.sell_on_graduation must be a boolean")
        result["sell_on_graduation"] = sell_on_graduation_raw

    return result


def _parse_limits(value: object) -> LimitsMsg:
    obj = _as_record(value, "limits")

    max_wallets_per_session = _optional_int(
        obj.get("max_wallets_per_session"),
        "limits.max_wallets_per_session",
    )
    max_positions_per_wallet = _optional_int(
        obj.get("max_positions_per_wallet"),
        "limits.max_positions_per_wallet",
    )
    max_sessions_per_api_key = _optional_int(
        obj.get("max_sessions_per_api_key"),
        "limits.max_sessions_per_api_key",
    )

    return {
        "hi_capacity": _as_int(obj.get("hi_capacity"), "limits.hi_capacity"),
        "pnl_flush_ms": _as_int(obj.get("pnl_flush_ms"), "limits.pnl_flush_ms"),
        "max_positions_per_session": _as_int(
            obj.get("max_positions_per_session"),
            "limits.max_positions_per_session",
        ),
        "max_wallets_per_session": 0 if max_wallets_per_session is None else max_wallets_per_session,
        "max_positions_per_wallet": 0 if max_positions_per_wallet is None else max_positions_per_wallet,
        "max_sessions_per_api_key": 0
        if max_sessions_per_api_key is None
        else max_sessions_per_api_key,
    }


def _set_optional_market_context(
    obj: dict[str, object],
    key: str,
    value: object,
    path: str,
) -> None:
    parsed = _optional_market_context(value, path)
    if parsed is not None:
        obj[key] = parsed


def _optional_market_context(value: object, path: str) -> MarketContextMsg | None:
    if value is None:
        return None
    return _parse_market_context(value, path)


def _parse_market_context(value: object, path: str) -> MarketContextMsg:
    obj = _as_record(value, path)

    result: MarketContextMsg = {
        "market_type": _parse_market_type(
            obj.get("market_type"),
            f"{path}.market_type",
        )
    }

    _set_optional_object(result, "pumpfun", obj.get("pumpfun"), f"{path}.pumpfun")
    _set_optional_object(result, "pumpswap", obj.get("pumpswap"), f"{path}.pumpswap")
    _set_optional_object(result, "meteora_dbc", obj.get("meteora_dbc"), f"{path}.meteora_dbc")
    _set_optional_object(
        result,
        "meteora_damm_v2",
        obj.get("meteora_damm_v2"),
        f"{path}.meteora_damm_v2",
    )
    _set_optional_object(
        result,
        "raydium_launchpad",
        obj.get("raydium_launchpad"),
        f"{path}.raydium_launchpad",
    )
    _set_optional_object(
        result,
        "raydium_cpmm",
        obj.get("raydium_cpmm"),
        f"{path}.raydium_cpmm",
    )

    return result


def _parse_market_type(value: object, path: str) -> MarketTypeMsg:
    market_type = _as_string(value, path)
    allowed_market_types: set[str] = {
        "pump_fun",
        "pump_swap",
        "meteora_dbc",
        "meteora_damm_v2",
        "raydium_launchpad",
        "raydium_cpmm",
    }
    if market_type not in allowed_market_types:
        raise ValueError(f"invalid market type at {path}: {market_type}")
    return cast(MarketTypeMsg, market_type)


def _set_optional_int(
    obj: dict[str, object],
    key: str,
    value: object,
    path: str,
) -> None:
    parsed = _optional_int(value, path)
    if parsed is not None:
        obj[key] = parsed


def _set_optional_str(
    obj: dict[str, object],
    key: str,
    value: object,
    path: str,
) -> None:
    parsed = _optional_string(value, path)
    if parsed is not None:
        obj[key] = parsed


def _set_optional_object(
    obj: dict[str, object],
    key: str,
    value: object,
    path: str,
) -> None:
    parsed = _optional_object(value, path)
    if parsed is not None:
        obj[key] = parsed


def _optional_string(value: object, path: str) -> str | None:
    if value is None:
        return None
    return _as_string(value, path)


def _optional_int(value: object, path: str) -> int | None:
    if value is None:
        return None
    return _as_int(value, path)


def _optional_object(value: object, path: str) -> dict[str, object] | None:
    if value is None:
        return None
    return _as_record(value, path)


def _as_record(value: object, path: str) -> dict[str, object]:
    if not isinstance(value, dict):
        raise ValueError(f"{path} must be an object")
    return cast(dict[str, object], value)


def _as_string(value: object, path: str) -> str:
    if not isinstance(value, str):
        raise ValueError(f"{path} must be a string")
    return value


def _as_float(value: object, path: str) -> float:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        raise ValueError(f"{path} must be a finite number")

    number = float(value)
    if not math.isfinite(number):
        raise ValueError(f"{path} must be a finite number")
    return number


def _as_int(value: object, path: str) -> int:
    if isinstance(value, bool):
        raise ValueError(f"{path} must be an integer")

    if isinstance(value, int):
        return value

    if isinstance(value, float) and value.is_integer() and math.isfinite(value):
        return int(value)

    raise ValueError(f"{path} must be an integer")


__all__ = [
    "BalanceUpdateServerMessage",
    "ClientMessage",
    "ClosePositionClientMessage",
    "ConfigureClientMessage",
    "ErrorServerMessage",
    "ExitSignalWithTxServerMessage",
    "HelloOkServerMessage",
    "LimitsMsg",
    "MarketContextMsg",
    "MarketTypeMsg",
    "PnlUpdateServerMessage",
    "PositionClosedServerMessage",
    "PositionOpenedServerMessage",
    "RequestExitSignalClientMessage",
    "ServerMessage",
    "StrategyConfigMsg",
    "UpdateStrategyClientMessage",
    "UpdateWalletsClientMessage",
    "client_message_from_obj",
    "client_message_from_text",
    "client_message_to_text",
    "server_message_from_obj",
    "server_message_from_text",
    "server_message_to_text",
]
